Rewrite Qiskit C API extension-module guide (#4890)
Qiskit/documentation: docs/guides/c-extension-for-python.mdx

# Extend Qiskit in Python with C

The Qiskit C API can be used within Python extension modules.
You can write performance-critical sections of your Qiskit extensions in C to accelerate them, and
then safely distribute these to your users.

This guide will walk you through the process of defining a complete extension module, configuring

> its build process, and exposing it to Python users.  The package will provide a simple port of

Reviewer: ```suggestion
its build process, and exposing it to Python users.  The package provides a simple port of
```

<Admonition type="tip">
  You may find the following external resources helpful:

  - [The CPython documentation on writing extension modules](https://docs.python.org/3/extending/index.html).
  - [The NumPy documentation on using its C API](https://numpy.org/doc/stable/user/c-info.how-to-extend.html).

  The Qiskit C API is exposed for Python extension modules in a very similar manner to the NumPy C
  API.  If you have previously programmed a NumPy extension, you will find the process on Qiskit
  familiar.
</Admonition>

<Admonition type="warning">
  The Qiskit C API is still experimental, and has not yet committed to a fully stable programming or 
  binary interface, and there may be breaking changes between minor versions.

  For example, an extension module using Qiskit v2.4.0 at build time is guaranteed to work with
  Qiskit v2.4.1 at runtime, but may break when using Qiskit v2.5.0 at runtime.
</Admonition>

## Requirements

We will start from a clean directory.

You must have the standard C compiler toolchain available for your platform.  You must also have a

> version of Python that includes its C API headers (this is standard).

Reviewer 2: Do you want to link to the "install the C API guide" which has a section on C requirements: https://quantum.cloud.ibm.com/docs/en/guides/install-c-api#requirements ?

> You should be familiar with, or prepared to look up, the individual functions and objects available

Reviewer 2: This part doesn't use the C API objects yet, right? If that only comes further below, where you're also explicitly referring to the C API docs I would suggest scratching that sentence (which also reads a little discouraging to me, but maybe that's just me 😛)

```toml
# file: pyproject.toml
[build-system]
requires = [
    "setuptools",
    "qiskit~=2.4.0",
]
build-backend = "setuptools.build_meta"

[project]
name = "spectator_measures"
authors = [
    { name = "Qiskit Developer" },
]
version = "0.0.1"
dependencies = [
    "qiskit~=2.4.0",
]
# If you intend to release your package, you should
# also set the `license` information, and so on.

[tool.setuptools]
package-dir = {"" = "src"}
```

As of Qiskit 2.4, the C API is not yet stable outside of minor versions (for example, v2.4.0 will be
compatible with v2.4.1 but not v2.5.0 in terms of the C API). In the future, we intend to extend
this stability to with major versions.  In the meantime, we set our runtime version of Qiskit in
`project.dependencies` to match the minor version used at build time.

In many pure-Python `setuptools`-based projects, it would be sufficient to simply have the
`pyproject.toml` file.  However, our module will need access to the Qiskit C API header files during
its build process.   Starting from v2.4, these are included in the Python distributions of the
Qiskit SDK, and you can locate the directory containing them using `qiskit.capi.get_include()`.

> This leads us to a `setup.py` file that looks like:

Reviewer: ```suggestion
This returns a `setup.py` file that looks like:
```

```python
# file: setup.py
import qiskit
from setuptools import setup, Extension

core_ext = Extension(
    # The fully qualified module name of the extension.
    name="spectator_measures._core",
    # The C source files needed for the extension.  The file
    # name is conventionally `<mod>module.c`, where `<mod>`
    # is the module name (`_core`, in this case).
    sources=["src/spectator_measures/_coremodule.c"],
    # Directories containing additional header files used in
    # the build process.
    include_dirs=[qiskit.capi.get_include()],
)
setup(ext_modules=[core_ext])
```

Most of the package information is defined in `pyproject.toml`, and `setuptools.setup()` will
also read that file.

<Admonition type="tip">
See the [`setuptools` User Guide](https://setuptools.pypa.io/en/latest/userguide/) for more
information on configuring `setuptools`-based projects.
</Admonition>

## Write the Python-space wrapper

It is technically possible to define everything in a Python extension from C.  In practice, it is
far easier to interact with other Python-space code from Python itself.

> Our package will define a custom transpiler pass that derives from the Python-space

Reviewer: ```suggestion
The Qiskit C package defines a custom transpiler pass that derives from the Python-space
```

   arguments, keyword arguments, and the functions to use to convert them.  If this fails, the
   function propagates the error.
1. Delegates to the C-native business logic of the previous section, which mutates the DAG in place.
1. Returns the Python-space `None` object.

The most complex logic is all inside `PyArg_ParseTupleAndKeywords`.  This is well documented in [the
CPython documentation on parsing arguments](https://docs.python.org/3/c-api/arg.html), which you
should consult for further information.

The Qiskit C API provides several functions with names like `qk_*_convert_from_python`, which are
designed as "converter" functions for use with `PyArg_Parse*`functions.  These correspond to the
`O&` keys in the format string; here, we used `qk_dag_convert_from_python` and
`qk_target_convert_from_python`.  These functions _borrow_ the C-native object from the Python
argument they derive from.  This means that mutations will propagate to Python space, but also that
you should take care not to release your reference to the Python object that backs them, while using
the result.  This is standard for Python C API programming.

> Next, we define the information about our module and the function it contains, so we can pass it to

Reviewer: ```suggestion
Next, we define the information about the Qiskit C module and the function it contains, so we can pass it to
```

`PyInit_<mod>`, where `<mod>` is the (unqualified) name of the module.  In this case, the fully
qualified module name is `spectator_measures._core`, and the unqualified name is `_core`, so our
function must be called `PyInit__core`, with the double underscore.

```c
# file: src/spectator_measures/_coremodule.c
PyMODINIT_FUNC PyInit__core(void) {
    // This line is critical to use the Qiskit C API. Your code will
    // likely be immediately terminated by the operating system if you
    // forget to do this.
    if (qk_import() < 0) {
        return NULL;
    };
    // The standard Python call to initialize a module.
    return PyModuleDef_Init(&core_module);
}
```

The `PyMODINIT_FUNC` and `PyModuleDef_Init` symbols are both standard Python C API programming.  The
Qiskit-specific component is `qk_import()`.  It is critical that you call this function during the
initialization function of your module; you will not be able to call any Qiskit C API functions
until this has been successfully executed.

## Use the package from Python

We now have a complete package, including a C extension module.  As we have used only standard

> tooling, and we do not require linking against any non-standard system libraries during build time,

Author: I was wrong - two more instances of increased passivity ;)
